Pagiante and ordering seems to be conflicting

When I declare `ordering` and `paginate_by` on a view I get `AssertionError`

`AssertionError: Cannot reorder a query once a slice has been taken.`

Issue is with line 280 in `helpers.py`
```python
elif issubclass(type(self.model_field), ForeignKey):
    pks = self.object_list.order_by().distinct().values_list('%s__pk' % self.field_name, flat=True)
```

--- a/smart_lists/helpers.py
+++ b/smart_lists/helpers.py
@@ -256,7 +256,7 @@ class SmartFilterValue(QueryParamsMixin, object):
 
 
 class SmartFilter(TitleFromModelFieldMixin, object):
-    def __init__(self, model, field, query_params, object_list):
+    def __init__(self, model, field, query_params, object_list, view):
         self.model = model
 
         if isinstance(field, SmartListFilter):
@@ -267,6 +267,7 @@ class SmartFilter(TitleFromModelFieldMixin, object):
             self.model_field = self.model._meta.get_field(self.field_name)
         self.query_params = query_params
         self.object_list = object_list
+        self.view = view
 
     def get_title(self):
         if isinstance(self.model_field, SmartListFilter):
@@ -291,7 +292,8 @@ class SmartFilter(TitleFromModelFieldMixin, object):
                 for choice in ((1, _('Yes')), (0, _('No')))
             ]
         elif issubclass(type(self.model_field), ForeignKey):
-            pks = self.object_list.order_by().distinct().values_list('%s__pk' % self.field_name, flat=True)
+            # use `self.view.object_list` in order to create filter from all objects not a paginated subset
+            pks = self.view.object_list.order_by().distinct().values_list('%s__pk' % self.field_name, flat=True)
             remote_field = self.model_field.rel if hasattr(self.model_field, 'rel') else self.model_field.remote_field
             qs = remote_field.model.objects.filter(pk__in=pks)
             values = [SmartFilterValue(self.field_name, obj, str(obj.pk), self.query_params) for obj in qs]
@@ -309,6 +311,7 @@ class SmartList(object):
         list_search=None,
         search_query_param=None,
         ordering_query_param=None,
+        view=None,
     ):
         self.object_list = object_list
         self.model = object_list.model
@@ -325,7 +328,7 @@ class SmartList(object):
 
         self.filters = (
             [
-                SmartFilter(self.model, field, self.query_params, self.object_list)
+                SmartFilter(self.model, field, self.query_params, self.object_list, view)
                 for i, field in enumerate(self.list_filter, start=1)
             ]
             if self.list_filter

--- a/smart_lists/templatetags/smart_list.py
+++ b/smart_lists/templatetags/smart_list.py
@@ -58,6 +58,7 @@ def smart_list(
         list_search=list_search,
         search_query_param=search_query_param,
         ordering_query_param=ordering_query_param,
+        view=context['view'],
     )
 
     split_grid_small_size = int(round(grid_size * 0.25))

--- a/testproject/management/commands/seed_data.py
+++ b/testproject/management/commands/seed_data.py
@@ -1,6 +1,11 @@
 from django.core.management import BaseCommand
 
-from testproject.models import SampleModel, CATEGORY_CHOICES
+from testproject.models import (
+    CATEGORY_CHOICES,
+    SampleModel,
+    ForeignModelWithoutUrl,
+    ForeignModelWithUrl,
+)
 
 
 class Command(BaseCommand):
@@ -8,5 +13,9 @@ class Command(BaseCommand):
         for i, (category, _) in enumerate(CATEGORY_CHOICES):
             for y in range(5):
                 title = "Example title {}{}".format(i, y)
-                SampleModel.objects.get_or_create(title=title, category=category)
+                f1_title = "ForeignModelWithUrl title {}{}".format(i, y)
+                f2_title = "ForeignModelWithoutUrl title {}{}".format(i, y)
+                f1, _ = ForeignModelWithUrl.objects.get_or_create(title=f1_title)
+                f2, _ = ForeignModelWithoutUrl.objects.get_or_create(title=f2_title)
+                SampleModel.objects.get_or_create(title=title, category=category, foreign_1=f1, foreign_2=f2)
         self.stdout.write("Example data created successfully.")

--- a/testproject/models.py
+++ b/testproject/models.py
@@ -17,9 +17,15 @@ class ForeignModelWithUrl(models.Model):
             args=(self.id,),
         )
 
+    def __str__(self):
+        return self.title
+
 
 class ForeignModelWithoutUrl(models.Model):
     title = models.CharField(max_length=128)
+
+    def __str__(self):
+        return self.title
 
 
 class SampleModel(models.Model):

--- a/testproject/urls.py
+++ b/testproject/urls.py
@@ -20,6 +20,6 @@ from testproject import views
 
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
-    url(r'', views.TestListView.as_view()),
+    url(r'foreign/', views.TestListView.as_view()),
     url(r'', views.SampleModelListView.as_view()),
 ]

--- a/testproject/views.py
+++ b/testproject/views.py
@@ -15,15 +15,17 @@ def example_render_function(obj):
 class SampleModelListView(SmartListMixin, ListView):
     model = SampleModel
     paginate_by = 5
+    ordering = ['category']
     ordering_allowed_fields = ['title']
     list_display = (
         'title',
+        'foreign_1',
         'category',
         'some_display_method',
         (render_column_template('testproject/example_column_template.html'), 'Upper filter used for field'),
         (example_render_function, "Time"),
     )
-    list_filter = ("category",)
+    list_filter = ("category", 'foreign_1')
 
 
 class TestListView(SmartListMixin, ListView):

